### PR TITLE
Yield keyword

### LINKED ISSUE #98: Replace explicit continuations with `yield`
It is probably easier to explain handlers if instead of explicit continuations we have a `yield` keyword. A handler case which now looks like

    | #op x k => .... k e ....

would become

    | #op x => ..... yield e .....

### PR DESCRIPTION
**Pull request by SkySkimmer**
Closes #98.

### CODE REVIEW

File: src/nucleus/eval.ml [hunk 1]
```diff
@@ -352,6 +352,12 @@ let rec infer env (c',loc) =
     let j = Judgement.mk_term ctx te ty in
     Value.return_term j
 
+  | Syntax.Yield ->
+    begin match Environment.lookup_continuation env with
+      | Some y -> Value.return y
+      | None -> Error.impossible ~loc "yield without continuation set"
```
Comment (andrejbauer) [inline src/nucleus/eval.ml]: Shouldn't this be a runtime error? It's possible for user to use `yield` outside a handler.

Comment (SkySkimmer) [inline src/nucleus/eval.ml]: We check that in desugar.ml (line 432+)

**Comment by SkySkimmer** [comment]
`simple_term` just means it doesn't need parenthesising in any position. We don't expect `(fun x => x)` to evaluate to type theory, but it's still going to be a `simple_term`.
There is a question of when we should require that something be applied (eg whnf) or not (yield in this PR) or even an external.
We do something special with whnf and refl in checking mode, so that could be one answer.

**Comment by andrejbauer** [comment]
Let us take `whnf` as an example. If we require it to be applied, then it is easy to evaluate. If we do not, then `whnf` has to evaluate to a closure. Thus we end up making a lot of closures all the time for no good reason because unapplied `whnf` is hardly ever useful, and if someone needs that they can η-expand it anyhow.

With tags the situation is different: a tag must always be fully applied or else we do not know how many arguments to expect. So we do not know what sort of closure to make when we see a bare tag.

With `yield` the situation is different because `yields` *is* a closure already. If we implemented continuations as something other than closures then we would have to rethink the situation.

In general, it is wrong to consider these questions to be about parsing. They are about evaluation. Parsing should be a consequence.

**Pull request merged**